editor/assets/default_renderpipeline/builtin-pipeline.ts

```diff
@@ -22,37 +22,38 @@
  THE SOFTWARE.
 */
 
-import { DEBUG } from 'internal:constants';
-import { sys, Vec2, Vec3, Vec4 } from '../../core';
-import { AABB } from '../../core/geometry/aabb';
-import { Frustum } from '../../core/geometry/frustum';
-import intersect from '../../core/geometry/intersect';
-import { Sphere } from '../../core/geometry/sphere';
-import { ClearFlagBit, Color, Format, FormatFeatureBit, LoadOp, StoreOp, Viewport } from '../../gfx/base/define';
-import { RenderScene } from '../../render-scene/core/render-scene';
-import { RenderWindow } from '../../render-scene/core/render-window';
-import { Camera, CameraUsage } from '../../render-scene/scene/camera';
-import { DirectionalLight } from '../../render-scene/scene/directional-light';
-import { Light, LightType } from '../../render-scene/scene/light';
-import { CSMLevel } from '../../render-scene/scene/shadows';
-import { SpotLight } from '../../render-scene/scene/spot-light';
-import { BasicPipeline, BasicRenderPassBuilder, PipelineBuilder } from './pipeline';
-import { QueueHint, SceneFlags } from './types';
-import { supportsR32FloatTexture } from '../define';
-import { Material } from '../../asset/assets';
-import { PipelineSettings } from './settings';
-import { getEditorPipelineCamera, getEditorPipelineSettings } from './framework';
-
-function forwardNeedClearColor (camera: Camera): boolean {
+import { DEBUG } from 'cc/env';
+import {
+    geometry,
+    gfx,
+    Material,
+    pipeline,
+    PipelineSettings,
+    renderer,
+    rendering,
+    sys,
+    Vec2,
+    Vec3,
+    Vec4,
+} from 'cc';
+
+const { AABB, Sphere, intersect } = geometry;
+const { ClearFlagBit, Color, Format, FormatFeatureBit, LoadOp, StoreOp, Viewport } = gfx;
+const { scene } = renderer;
+const { QueueHint, SceneFlags } = rendering ? rendering
+    : { QueueHint: undefined, SceneFlags: undefined };
```

Comment:
`rendering` will be `undefined` if it is feature cropped.
As a result, destructuring `undefined` will cause runtime error.
Here we provide a backup object as workaround.